internal/content_coding.go

```diff
@@ -1,18 +1,115 @@
 package internal
 
 import (
+	"bufio"
 	"bytes"
 	"compress/gzip"
 	"errors"
 	"io"
 )
 
+// NewStreamContentEncoder returns a writer that will encode the stream
+// according to the encoding type.
+func NewStreamContentEncoder(encoding string, w io.Writer) (io.Writer, error) {
+	switch encoding {
+	case "gzip":
+		return NewGzipWriter(w)
+	case "identity", "":
+		return w, nil
+	default:
+		return nil, errors.New("invalid value for content_encoding")
+	}
+}
+
+// NewStreamContentDecoder returns a reader that will decode the stream
+// according to the encoding type.
+func NewStreamContentDecoder(encoding string, r io.Reader) (io.Reader, error) {
+	switch encoding {
+	case "gzip":
+		return NewGzipReader(r)
+	case "identity", "":
+		return r, nil
+	default:
+		return nil, errors.New("invalid value for content_encoding")
+	}
+}
+
+// GzipWriter is similar to the gzip.Writer but flushes after each call to
+// write, this ensure that the written data can be decompressed on the read
+// size immediately.
+type GzipWriter struct {
+	w io.Writer
+	z *gzip.Writer
+}
+
+func NewGzipWriter(w io.Writer) (io.Writer, error) {
+	return &GzipWriter{w: w, z: new(gzip.Writer)}, nil
+}
+
+func (w *GzipWriter) Write(p []byte) (int, error) {
+	w.z.Reset(w.w)
```

Comment:
I'm not sure why we start a new gzip stream for each write.  Doesn't this prevent the gzip algorithm from getting enough state to compress well?

Reply:
It would if the write is small, my idea here was that it would be nice to be able to read each batch.  I'm going to remove the Writer completely for now since it isn't used and we can sort this out later.